### PR TITLE
net/dns/{., resolver}, wgengine: fix goroutine leak on shutdown

### PR DESCRIPTION
**Pull request by twitchyliquid64**

### CODE REVIEW

File: net/dns/manager.go [hunk 1]
```diff
@@ -34,6 +34,8 @@ var (
 
 var (
 	errFullQueue = errors.New("request queue full")
+	// ErrClosed is returned when the Manager has been shut down.
```
Comment (maisem) [inline net/dns/manager.go:37]: Use net.ErrClosed instead

File: net/dns/manager.go [hunk 2]
```diff
@@ -274,7 +284,13 @@ func (m *Manager) EnqueuePacket(bs []byte, proto ipproto.Proto, from, to netaddr
 //
 // TODO(tom): Rip out once all platforms use netstack.
 func (m *Manager) NextPacket() ([]byte, error) {
-	resp := <-m.responses
+	var resp response
+	select {
+	case <-m.ctx.Done():
+		return nil, ErrClosed
```
Comment (maisem) [inline net/dns/manager.go:290]: or ctx.Error()

File: net/dns/manager.go [hunk 3]
```diff
@@ -315,6 +331,13 @@ func (m *Manager) NextPacket() ([]byte, error) {
 }
 
 func (m *Manager) Query(ctx context.Context, bs []byte, from netaddr.IPPort) ([]byte, error) {
+	select {
+	case <-m.ctx.Done():
+		return nil, ErrClosed
```
Comment (maisem) [inline net/dns/manager.go:336]: Same

File: net/dns/resolver/tsdns.go [hunk 4]
```diff
@@ -264,7 +261,7 @@ func (r *Resolver) Query(ctx context.Context, bs []byte, from netaddr.IPPort) ([
 	select {
 	case <-r.closed:
 		metricDNSQueryErrorClosed.Add(1)
-		return nil, ErrClosed
+		return nil, errors.New("closed")
```
Comment (maisem) [inline net/dns/resolver/tsdns.go:264]: net.ErrClosed

File: wgengine/userspace.go [hunk 5]
```diff
@@ -499,7 +499,7 @@ func (e *userspaceEngine) handleLocalPackets(p *packet.Parsed, t *tstun.Wrapper)
 func (e *userspaceEngine) pollResolver() {
 	for {
 		bs, err := e.dns.NextPacket()
-		if err == resolver.ErrClosed {
+		if err == dns.ErrClosed {
```
Comment (maisem) [inline wgengine/userspace.go:502]: Same

File: net/dns/manager.go (hunk 2 above)
Comment (twitchyliquid64) [reply net/dns/manager.go:290]: This needs to match in `wgengine.pollResolver`, so ill leave it as `net.ErrClosed`

File: wgengine/userspace.go [hunk 6]
```diff
@@ -499,7 +500,7 @@ func (e *userspaceEngine) handleLocalPackets(p *packet.Parsed, t *tstun.Wrapper)
 func (e *userspaceEngine) pollResolver() {
 	for {
 		bs, err := e.dns.NextPacket()
-		if err == resolver.ErrClosed {
+		if err == net.ErrClosed {
```
Comment (maisem) [inline wgengine/userspace.go:503]: To be safer, use `errors.Is(err, net.ErrClosed)`

**Review by maisem — APPROVED** [review]

**Pull request merged**